### PR TITLE
Improvements for black-/whitelist scripts

### PR DESCRIPTION
**Pull request by DL6ER**
**By submitting this pull request, I confirm the following (please check boxes, eg [X])Failure to fill the template will close your PR:**

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [X] I have checked that [another pull request](https://github.com/pi-hole/pi-hole/pulls) for this purpose does not exist.
- [X] I have considered, and confirmed that this submission will be valuable to others.
- [X] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [X] I give this submission freely, and claim no ownership to its content.

**How familiar are you with the codebase?:**
- [X] 9

---
Fixes web interface bug [#116](https://github.com/pi-hole/AdminLTE/issues/116)

```
pi@raspberrypi:~# git checkout development
pi@raspberrypi:~# pihole -b "GitHUb.CoM"
::: GitHUb.CoM is not a valid argument or domain name

pi@raspberrypi:~# git checkout caseinsensitive
pi@raspberrypi:~# pihole -b "GitHUb.CoM"
::: Adding github.com to /etc/pihole/blacklist.txt...
::: github.com does not exist in /etc/pihole/whitelist.txt, no need to remove!
:::
::: Neutrino emissions detected...
:::
::: No custom adlist file detected, reading from default file... done!
:::
::: Using cached Event Horizon list...
::: 91279 unique domains trapped in the event horizon.
:::
::: Adding adlist sources to the whitelist... done!
::: Whitelisting 7 domains... done!
::: Blacklisting 1 domain... done!
::: Formatting domains into a HOSTS file... done!
:::
::: Cleaning up un-needed files... done!
:::
::: Refresh lists in dnsmasq... done!
::: Pi-hole blocking is Enabled
```

_This template was created based on the work of [`udemy-dl`](https://github.com/nishad/udemy-dl/blob/master/LICENSE)._

### CODE REVIEW

File: advanced/Scripts/list.sh [hunk 1]
```diff
@@ -55,7 +55,7 @@ HandleOther(){
 	domain=$(sed -e "y/ABCDEFGHIJKLMNOPQRSTUVWXYZ/abcdefghijklmnopqrstuvwxyz/" <<< "$1")
 
 	#check validity of domain
-	validDomain=$(echo "$domain" | perl -ne'print if /\b((?=[a-z0-9-]{1,63}\.)(xn--)?[a-z0-9]+(-[a-z0-9]+)*\.)+[a-z]{2,63}\b/')
```
Comment (PromoFaux) [inline advanced/Scripts/list.sh]: I may or may not have copy-pasted from a stackexchange post for this... ;)

**Comment by Mcat12** [comment]
Approved

**Comment by Mcat12** [comment]
Pending. It allows `test.com/ds*` into the whitelist.

**Comment by DL6ER** [comment]
Confirmed. The current `development` branch also allows adding this domain to black- and whitelist. My changes do not affect that. Here [old](https://github.com/pi-hole/pi-hole/pull/941/files#diff-1dcd206a83231abc1bb8d02aa72324c7L55) and [here](https://github.com/pi-hole/pi-hole/pull/941/files#diff-1dcd206a83231abc1bb8d02aa72324c7R58) we explicitly allow the character "*" in valid domain names.

Removing it fails since we would have to escape the asterisk as we use `grep` with exact line matching (`-x`) [here](https://github.com/pi-hole/pi-hole/pull/941/files#diff-1dcd206a83231abc1bb8d02aa72324c7R110). However, "*" has a different meaning in regexp than elsewhere, since here it is the `repetition operator` instead of the wildcard operator like in bash (as e.g. in `ls *.log`).

I'll make the necessary changes and commit them right here.

**Comment by DL6ER** [comment]
<pre>
pi@raspberry:~# sudo pihole -w text.com/ds*
::: Adding text.com/ds* to /etc/pihole/whitelist.txt...
::: text.com/ds* does not exist in /etc/pihole/blacklist.txt, no need to remove!
:::
::: Neutrino emissions detected...
:::
::: No custom adlist file detected, reading from default file... done!
:::
::: Using cached Event Horizon list...
::: 105365 unique domains trapped in the event horizon.
:::
::: Adding adlist sources to the whitelist... done!
::: Whitelisting 19 domains... done!
::: Blacklisting 0 domains... done!
::: Formatting domains into a HOSTS file... done!
:::
::: Cleaning up un-needed files... done!
:::
::: Refresh lists in dnsmasq... done!
::: Pi-hole blocking is Enabled
pi@raspberry:~# sudo pihole -w text.com/ds*
::: text.com/ds* already exists in /etc/pihole/whitelist.txt, no need to add!
::: text.com/ds* does not exist in /etc/pihole/blacklist.txt, no need to remove!
pi@raspberry:~# sudo pihole -w -d text.com/ds*
::: Removing text.com/ds* from /etc/pihole/whitelist.txt...
:::
::: Neutrino emissions detected...
:::
::: No custom adlist file detected, reading from default file... done!
:::
::: Using cached Event Horizon list...
::: 105365 unique domains trapped in the event horizon.
:::
::: Adding adlist sources to the whitelist... done!
::: Whitelisting 18 domains... done!
::: Blacklisting 0 domains... done!
::: Formatting domains into a HOSTS file... done!
:::
::: Cleaning up un-needed files... done!
:::
::: Refresh lists in dnsmasq... done!
::: Pi-hole blocking is Enabled
pi@raspberry:~# sudo pihole -w -d text.com/ds*
::: text.com/ds* does not exist in /etc/pihole/whitelist.txt, no need to remove!
</pre>

**Comment by Mcat12** [comment]
Are you able to delete that domain using this code? That was something that wasn't working (it said it wasn't in the whitelist).

**Comment by DL6ER** [comment]
@Mcat12 Yes, see [here](https://github.com/pi-hole/pi-hole/pull/941#issuecomment-264439086).

**Comment by Mcat12** [comment]
Oh, was that using the new code or old code? I'll retest soon.

**Comment by DL6ER** [comment]
The current version of the code in this PR (commit https://github.com/pi-hole/pi-hole/pull/941/commits/c8ad6f23a88f1e292796f508dd3cafc99c5765de)

File: advanced/Scripts/list.sh [hunk 2]
```diff
@@ -50,9 +50,18 @@ EOM
 	exit 1
 }
 
+EscapeRegexp() {
+    # This way we may safely insert an arbitrary
+    # string in our regular expressions
+    echo $* | sed 's/[]\.|$(){}?+*^]/\\&/g' | sed 's/\//\\\//g'
```
Comment (DL6ER) [inline advanced/Scripts/list.sh]: We have to force this one. I'm not even sure if this particular `sed` can actually be done with double quotes. There is nothing to be expanded here, so codacy's comment doesn't matter here.

**Comment by Mcat12** [comment]
Still not able to remove the domain for some reason...
Image

**Comment by DL6ER** [comment]
@Mcat12 I realized that the simplification I did in my recent commit (509cfd1) works on my local machine but not on my Raspberry. Interestingly... I reverted it in 78cb43d. Please try again.

**Comment by Mcat12** [comment]
Works now.

**Comment by Mcat12** [comment]
Approved

**Pull request merged**